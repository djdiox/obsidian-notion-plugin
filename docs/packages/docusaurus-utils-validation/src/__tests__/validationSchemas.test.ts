/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Joi from '../Joi';

import {
  AdmonitionsSchema,
  RehypePluginsSchema,
  RemarkPluginsSchema,
  PluginIdSchema,
  URISchema,
  PathnameSchema,
} from '../validationSchemas';

function createTestHelpers({
  schema,
  defaultValue,
}: {
  schema: Joi.Schema;
  defaultValue?: unknown;
}) {
  function testOK(value: unknown) {
    expect(Joi.attempt(value, schema)).toEqual(value ?? defaultValue);
  }

  function testFail(value: unknown) {
    expect(() => Joi.attempt(value, schema)).toThrowErrorMatchingSnapshot(
      `for value=${JSON.stringify(value)}`,
    );
  }

  return {testOK, testFail};
}

function testMarkdownPluginSchemas(schema: Joi.Schema) {
  const {testOK, testFail} = createTestHelpers({
    schema,
    defaultValue: [],
  });

  testOK(undefined);
  testOK([() => {}]);
  testOK([[() => {}, {attr: 'val'}]]);
  testOK([[() => {}, {attr: 'val'}], () => {}, [() => {}, {attr: 'val'}]]);
  // cSpell:ignore remarkjs
  // official `remarkjs/remark-frontmatter` plugin accepts string options
  testOK([[() => {}, 'string-option']]);
  testOK([[() => {}, true]]);

  testFail(null);
  testFail(false);
  testFail(3);
  testFail([null]);
  testFail([false]);
  testFail([3]);
  testFail([[]]);
  testFail([[() => {}, undefined]]);
}

describe('validation schemas', () => {
  it('pluginIdSchema', () => {
    const {testOK, testFail} = createTestHelpers({
      schema: PluginIdSchema,
      defaultValue: 'default',
    });

    testOK(undefined);
    testOK('docs');
    testOK('default');
    testOK('plugin-id_with-simple-special-chars');
    testOK('doc1');

    testFail('/docs');
    testFail('docs/');
    testFail('do/cs');
    testFail('do cs');
    testFail(null);
    testFail(3);
    testFail(true);
    testFail([]);
  });

  it('admonitionsSchema', () => {
    const {testOK, testFail} = createTestHelpers({
      schema: AdmonitionsSchema,
      defaultValue: true,
    });

    testOK(undefined);
    testOK(true);
    testOK(false);
    testOK({});
    testOK({tag: '+++'});
    testOK({keywords: ['info', 'tip']});
    testOK({tag: '+++', keywords: ['info', 'tip']});

    testFail(3);
    testFail([]);
    testFail({unknownAttribute: 'val'});
    testFail({tag: ''});
    testFail({keywords: []});

    // Legacy types
    testFail({
      infima: true,
    });
    testFail({
      icons: 'emoji',
    });
    testFail({
      customTypes: {
        myKeyword: {
          keyword: `myKeyword`,
          infima: true,
          svg: '<svg width="512px" height="512px" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg"></svg>',
        },
      },
    });
  });

  it('remarkPluginsSchema', () => {
    testMarkdownPluginSchemas(RemarkPluginsSchema);
  });

  it('rehypePluginsSchema', () => {
    testMarkdownPluginSchemas(RehypePluginsSchema);
  });

  it('uRISchema', () => {
    const {testFail, testOK} = createTestHelpers({schema: URISchema});

    const validURL = 'https://docusaurus.io';
    const doubleHash = 'https://docusaurus.io#github#/:';
    const invalidURL = 'spaces are invalid in a URL';
    const relativeURL = 'relativeURL';
    const relativeURLWithParent = '../relativeURLWithParent';
    const urlFromIssue = 'https://riot.im/app/#/room/#ligo-public:matrix.org';
    testOK(validURL);
    testOK(doubleHash);
    testFail(invalidURL);
    testOK(relativeURL);
    testOK(relativeURLWithParent);
    testOK(urlFromIssue);

    const protocolRelativeUrl1 = '//docusaurus.io/path';
    const protocolRelativeUrl2 = '//docusaurus.io/docs/doc1#hash';
    testOK(protocolRelativeUrl1);
    testOK(protocolRelativeUrl2);
  });

  it('pathnameSchema', () => {
    const {testFail, testOK} = createTestHelpers({schema: PathnameSchema});

    testOK('/foo');
    testFail('foo');
    testFail('https://github.com/foo');
  });
});
