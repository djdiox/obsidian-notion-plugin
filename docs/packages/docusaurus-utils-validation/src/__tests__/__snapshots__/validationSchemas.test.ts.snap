// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation schemas admonitionsSchema: for value=[] 1`] = `""value" does not look like a valid admonitions config"`;

exports[`validation schemas admonitionsSchema: for value={"customTypes":{"myKeyword":{"keyword":"myKeyword","infima":true,"svg":"<svg width=\\"512px\\" height=\\"512px\\" viewBox=\\"0 0 512 512\\" xmlns=\\"http://www.w3.org/2000/svg\\"></svg>"}}} 1`] = `
"The Docusaurus admonitions system has changed, and the option "customTypes" does not exist anymore.
You now need to swizzle the admonitions component to provide UI customizations such as icons.
Please refer to https://github.com/facebook/docusaurus/pull/7152 for detailed upgrade instructions."
`;

exports[`validation schemas admonitionsSchema: for value={"icons":"emoji"} 1`] = `
"The Docusaurus admonitions system has changed, and the option "icons" does not exist anymore.
You now need to swizzle the admonitions component to provide UI customizations such as icons.
Please refer to https://github.com/facebook/docusaurus/pull/7152 for detailed upgrade instructions."
`;

exports[`validation schemas admonitionsSchema: for value={"infima":true} 1`] = `
"The Docusaurus admonitions system has changed, and the option "infima" does not exist anymore.
You now need to swizzle the admonitions component to provide UI customizations such as icons.
Please refer to https://github.com/facebook/docusaurus/pull/7152 for detailed upgrade instructions."
`;

exports[`validation schemas admonitionsSchema: for value={"keywords":[]} 1`] = `""keywords" does not contain 1 required value(s)"`;

exports[`validation schemas admonitionsSchema: for value={"tag":""} 1`] = `""tag" is not allowed to be empty"`;

exports[`validation schemas admonitionsSchema: for value={"unknownAttribute":"val"} 1`] = `""unknownAttribute" is not allowed"`;

exports[`validation schemas admonitionsSchema: for value=3 1`] = `""value" does not look like a valid admonitions config"`;

exports[`validation schemas pathnameSchema: for value="foo" 1`] = `""value" is not a valid pathname. Pathname should start with slash and not contain any domain or query string."`;

exports[`validation schemas pathnameSchema: for value="https://github.com/foo" 1`] = `""value" is not a valid pathname. Pathname should start with slash and not contain any domain or query string."`;

exports[`validation schemas pluginIdSchema: for value="/docs" 1`] = `"Illegal plugin ID value "/docs": it should only contain alphanumerics, underscores, and dashes."`;

exports[`validation schemas pluginIdSchema: for value="do cs" 1`] = `"Illegal plugin ID value "do cs": it should only contain alphanumerics, underscores, and dashes."`;

exports[`validation schemas pluginIdSchema: for value="do/cs" 1`] = `"Illegal plugin ID value "do/cs": it should only contain alphanumerics, underscores, and dashes."`;

exports[`validation schemas pluginIdSchema: for value="docs/" 1`] = `"Illegal plugin ID value "docs/": it should only contain alphanumerics, underscores, and dashes."`;

exports[`validation schemas pluginIdSchema: for value=[] 1`] = `""value" must be a string"`;

exports[`validation schemas pluginIdSchema: for value=3 1`] = `""value" must be a string"`;

exports[`validation schemas pluginIdSchema: for value=null 1`] = `""value" must be a string"`;

exports[`validation schemas pluginIdSchema: for value=true 1`] = `""value" must be a string"`;

exports[`validation schemas rehypePluginsSchema: for value=[[]] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas rehypePluginsSchema: for value=[[null,null]] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas rehypePluginsSchema: for value=[3] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas rehypePluginsSchema: for value=[false] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas rehypePluginsSchema: for value=[null] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas rehypePluginsSchema: for value=3 1`] = `""value" must be an array"`;

exports[`validation schemas rehypePluginsSchema: for value=false 1`] = `""value" must be an array"`;

exports[`validation schemas rehypePluginsSchema: for value=null 1`] = `""value" must be an array"`;

exports[`validation schemas remarkPluginsSchema: for value=[[]] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas remarkPluginsSchema: for value=[[null,null]] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas remarkPluginsSchema: for value=[3] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas remarkPluginsSchema: for value=[false] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas remarkPluginsSchema: for value=[null] 1`] = `
""[0]" does not look like a valid MDX plugin config. A plugin config entry should be one of:
- A tuple, like \`[require("rehype-katex"), { strict: false }]\`, or
- A simple module, like \`require("remark-math")\`"
`;

exports[`validation schemas remarkPluginsSchema: for value=3 1`] = `""value" must be an array"`;

exports[`validation schemas remarkPluginsSchema: for value=false 1`] = `""value" must be an array"`;

exports[`validation schemas remarkPluginsSchema: for value=null 1`] = `""value" must be an array"`;

exports[`validation schemas uRISchema: for value="spaces are invalid in a URL" 1`] = `""value" does not look like a valid url (value='')"`;
